src/leo_s3_auth.erl

```diff
@@ -460,16 +589,63 @@ get_auth_info() ->
             not_found
     end.
 
+%% @doc Construct Canonical Headers
+%% @private
+auth_v4_headers(Req, SignedHeaders) ->
+    HeaderList = binary:split(SignedHeaders, <<";">>, [global]),
+    auth_v4_headers(Req, HeaderList, <<>>).
+
+auth_v4_headers(_Req, [], Acc) ->
+    Acc;
+auth_v4_headers(Req, [Head|Rest], Acc) ->
+    Val = case cowboy_req:header(Head, Req) of
+              {undefined, _} -> <<>>;
+              {Bin, _}       -> Bin
+          end,
+    auth_v4_headers(Req, Rest, <<Acc/binary, Head/binary, ":", Val/binary, "\n">>).
+
+%% @doc Consutrct Canonical Query String
+%% @private
+auth_v4_qs(QueryStr) ->
+    List = binary:split(QueryStr, <<"&">>, [global]),
+    auth_v4_qs(List, "").
+
+auth_v4_qs([], Acc) ->
+    Acc;
+auth_v4_qs([<<>>|Rest], Acc) ->
+    auth_v4_qs(Rest, Acc);
+auth_v4_qs([Head|Rest], Acc) ->
+    {Key, Val} = case binary:match(Head, <<"=">>) of
+                     nomatch ->
+                         {Head, <<>>};
+                     _ ->
+                         [Key2, Val2] = binary:split(Head, <<"=">>),
+                         {Key2, Val2}
+                 end,
+    KeyEnc = http_uri:encode(binary_to_list(Key)),
+    ValEnc = http_uri:encode(binary_to_list(Val)),
```

Comment:
Use `cow_qs:urlencode/1` instead of `http_uri:encode` due to performance reason (no need to call binary_to_list)
https://github.com/ninenines/cowlib/blob/1.0.1/src/cow_qs.erl#L431